### ISSUE TITLE
Questions about the GRU structure

### ISSUE THREAD

**Issue by bright1993ff66** [ageron/handson-ml#132]
Thanks for your book. This book is amazing! But now I have some questions about the structure of the GRU cell is the Chapter 14 Recurrent Neural Network.  In this book, the structure of GRU you show is here:

![image](https://user-images.githubusercontent.com/29517359/33409632-5b1195f0-d5b7-11e7-9642-a794cc3d917e.png)

And you list the inner computations in this structure:

![image](https://user-images.githubusercontent.com/29517359/33409833-57a6279a-d5b8-11e7-8802-25700e857203.png)

My first question is:
does your picture miss the notation g(t)? Based on my understanding, g(t) should be the output after the running activation function tanh, am I right?
My second question is:
I don't quite understand the formula of h(t) there. From the picture, I don't see any more tanh in the picture. In my view, the h(t) should be computed as the following:

![image](https://user-images.githubusercontent.com/29517359/33410009-3814c98a-d5b9-11e7-87a5-c62cf71d14f3.png)

Am I right? Thank you very much!

**Comment by ageron** [comment]
Hi @bright1993ff66 ,

Thanks for your feedback, I very much appreciate it.

I'll add the missing g(t) on the diagram, good catch! 👍 

Regarding h(t), you are right, it was a bad copy/paste error in the first print of the book. Per Thorell noticed it and submitted [an errata in May](http://www.oreilly.com/catalog/errata.csp?isbn=0636920052289) (search for "GRU" on the errata page) so I fixed it. I'm guessing your copy was printed earlier, sorry about that.

Then in October, Nick Pogrebnyakov pointed out that the original paper actually computes:
h_t = (1 - z_t) * g_t + z_t * h_(t - 1)
In other words, the black (-1) on the diagram should be 1.5cm to the right.  This issue is not a big deal, because the GRU cell would work just as well: finding the right value of z to minimize the loss is pretty close to finding the right value of 1-z (the only difference is the starting point of training: when the model is initialized, its outputs are typically close to zero). Nonetheless, I fixed this error as well, so the latest printed versions are fine.

If you have one of the original copies of my book, then you probably should take a look at the other errata as well.  Most are minor, but there were a few more serious ones. The most serious problem was in the second half of chapter 16, in the Q-Learning section (the notebook gives all the details).

I sincerely apologize for the inconvenience.
Cheers,
Aurélien
